### PR TITLE
Minimize syzkaller testcases

### PR DESCRIPTION
**Pull request by kennethkang**
Rev 1:
* run minimize when test case reproduced successfully
* implement minimization
* unit test

### CODE REVIEW

File: src/clusterfuzz/_internal/bot/fuzzers/syzkaller/engine.py [hunk 1]
```diff
@@ -204,9 +217,17 @@ def minimize_corpus(self, target_path, arguments, input_dirs, output_dir,
...
     raise NotImplementedError
 
-  def minimize_testcase(self, target_path, arguments, input_path, output_path,
-                        max_time):
+  def minimize_testcase(
+      self,
+      target_path,  # pylint: disable=unused-argument
+      arguments,  # pylint: disable=unused-argument
+      input_path: str,
+      output_path,  # pylint: disable=unused-argument
```
Comment (oliverchang) [inline src/clusterfuzz/_internal/bot/fuzzers/syzkaller/engine.py:225]: nit: this should conform to the Engine interface in that `output_path` should be used.

**Review by oliverchang — COMMENTED** [review]
LGTM with some minor comments

File: src/clusterfuzz/_internal/bot/fuzzers/syzkaller/engine.py (hunk 1 above)
Comment (kennethkang) [reply src/clusterfuzz/_internal/bot/fuzzers/syzkaller/engine.py:225]: `syz-repro` (minimization tool) currently only takes in an input path to a testcase file from a crash.

File: src/clusterfuzz/_internal/bot/fuzzers/syzkaller/runner.py [hunk 2]
```diff
@@ -218,6 +218,24 @@ def repro(self, repro_timeout: int,
         output=result.output,
     )
 
+  def minimize(self, minimize_args: List[str]) -> engine.MinimizeResult:
+    """Minimizing crash testcase.
+    Args:
+      minimize_args: list of arguments to be passed to syz-repro.
+    """
+    logs.log('Running Syzkaller Minimization (syz-repro) against testcase.')
+    additional_args = copy.copy(minimize_args)
```
[the remark this answers is not in the document]
Comment (kennethkang) [reply src/clusterfuzz/_internal/bot/fuzzers/syzkaller/runner.py:227]: 👍

File: src/clusterfuzz/fuzz/engine.py [hunk 3]
```diff
@@ -53,7 +53,17 @@ def __init__(self, logs, command, crashes, stats, time_executed):
...
+  """Results from running reproce (syz-crush) against a testcase."""
+
+  def __init__(self, command, return_code, time_executed, output):
+    self.command = command
+    self.return_code = return_code
+    self.time_executed = time_executed
+    self.output = output
+
+
+class MinimizeResult(object):
```
[the remark this answers is not in the document]
Comment (kennethkang) [reply src/clusterfuzz/fuzz/engine.py:65]: 👍

File: src/clusterfuzz/_internal/bot/fuzzers/syzkaller/engine.py (hunk 1 above)
Comment (kennethkang) [reply src/clusterfuzz/_internal/bot/fuzzers/syzkaller/engine.py:225]: hm I see what you mean. Currently `syz-repro` does not produce any output. However, I currently have an open PR for `syzkaller` that does produce an output from this tool. I can reflect this change and open another PR after this once the other `syzkaller` PR is merged in. Then we can use this `output path` and perhaps we can upstream this minimized output to a bucket or something.

Comment (kennethkang) [reply src/clusterfuzz/_internal/bot/fuzzers/syzkaller/engine.py:225]: https://github.com/google/syzkaller/pull/2955

Comment (oliverchang) [reply src/clusterfuzz/_internal/bot/fuzzers/syzkaller/engine.py:225]: We can still support the correct interface semantics here, by:

- Copying `input_path` to `output_path`.
- Passing `output_path` to syzkaller.

**Review by oliverchang — APPROVED** [review]

**Pull request merged**